includes/class-newspack-newsletters-renderer.php

```diff
@@ -558,6 +554,42 @@ private static function render_mjml( $post ) {
 			}
 		}
 
+		// Insert any ads.
+		if ( $include_ads ) {
+			$ads_query  = new WP_Query(
+				array(
+					'post_type'      => Newspack_Newsletters_Ads::NEWSPACK_NEWSLETTERS_ADS_CPT,
+					'posts_per_page' => -1,
+				)
+			);
+			$ads        = $ads_query->get_posts();
+			$ads_markup = '';
+			foreach ( $ads as $ad ) {
+				$expiry_date = new DateTime( get_post_meta( $ad->ID, 'expiry_date', true ) );
+				$now_date    = new DateTime();
+				if ( $expiry_date > $now_date ) {
+					$ads_markup .= self::post_to_mjml_components( $ad, false );
+				}
+			}
```

Comment:
You could use [the meta query options for `WP_Query`](https://developer.wordpress.org/reference/classes/wp_query/#custom-field-post-meta-parameters) to make it so that the WP_Query only fetches valid ads, and then we don't have to loop through them all checking if they're valid. The query would only fetch newsletter ad posts with `expiry_date` meta greater-than-or-equal-to the current date. I'm not sure whether there's a good way to do that currently since the meta values look like `2020-06-29T11:37:00`, but it would be trivial if the meta were stored as unix timestamps  because it could just compare the integer values of the expiry timestamp to the timestamp from the current day at `00:00:00`.

Something like this is what I'm trying to explain here. :) 
```
array(
    'post_type'      => Newspack_Newsletters_Ads::NEWSPACK_NEWSLETTERS_ADS_CPT,
    'posts_per_page' => -1,
    'meta_key' => 'expiry_date',
    'meta_value_num' => $start_of_today,
    'meta_compare' => '>=',
)

```

I don't think the current implementation will cause any issues since there won't be thousands of newsletter ads to iterate through, but just a potential improvement.

Reply:
Fixed in 3bbc88a – and also ensured that in both editor and on the backend the expiry date is handled in the same way – comparing full days.

Reply:
Good solution! I didn't realize we could do meta queries by date. :)